lib/Dialect/Triton/Transforms/LoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,90 @@
+#include "mlir/Transforms/LoopInvariantCodeMotionUtils.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+#include "triton/Dialect/Triton/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/Transforms/PipeliningUtility.h"
+#include "llvm/Support/Debug.h"
+
+#define GEN_PASS_CLASSES
+#include "triton/Dialect/Triton/Transforms/Passes.h.inc"
+
+#define DEBUG_TYPE "triton-licm"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace mlir::triton {
+
+namespace {
+
+class LoopInvariantCodeMotionPass
+    : public TritonLoopInvariantCodeMotionBase<LoopInvariantCodeMotionPass> {
+
+  DenseMap<LoopLikeOpInterface, bool> isLoopMemoryEffectFreeOrOnlyRead;
+
+  bool isMemoryEffectFreeOrOnlyRead(Operation *op) {
+    std::optional<SmallVector<MemoryEffects::EffectInstance>> effects =
+        getEffectsRecursively(op);
+    if (!effects)
+      return false;
+    return llvm::all_of(*effects,
+                        [&](const MemoryEffects::EffectInstance &effect) {
+                          return isa<MemoryEffects::Read>(effect.getEffect());
+                        });
+  }
+
+  void runOnOperation() override {
+    // Walk through all loops in a function in innermost-loop-first order.
+    // This way, we first LICM from the inner loop, and place the ops in the
+    // outer loop, which in turn can be further LICM'ed.
+    getOperation()->walk([&](LoopLikeOpInterface loopLike) {
+      moveLoopInvariantCode(
+          loopLike.getLoopRegions(),
+          // isDefinedOutsideOfRegion
+          [&](Value value, Region *region) {
+            return loopLike.isDefinedOutsideOfLoop(value);
+          },
+          // shouldMoveOutOfRegion
+          [&](Operation *op, Region *region) {
+            if (!isa<LoadOp>(op))
+              return isSpeculatable(op) && isMemoryEffectFree(op);
+            if (!isLoopMemoryEffectFreeOrOnlyRead.contains(loopLike))
+              isLoopMemoryEffectFreeOrOnlyRead[loopLike] =
+                  isMemoryEffectFreeOrOnlyRead(loopLike);
+            return isMemoryEffectFreeOrOnlyRead(op) &&
```

Comment:
This seems wrong, we can only move `LoadOp` specifically, not any generic op that only reads. I'm thinking specifically descriptor load may hit this case.

Reply:
Oh looking again you have an early exit for !isa<LoadOp> so this check is just to exclude volatile loads, makes sense.